File: recursos/codigo/verifica_ambiente.py
"""
Verificação do ambiente da disciplina DPW — material de apoio do M00.

Confere as DUAS camadas: Python/Django/DRF (backend) e Node/pnpm (frontend).

Rode com o ambiente virtual ativo:

    python verifica_ambiente.py

Saída: um relatório com OK/FALHA por item e a dica de correção quando falhar.
Código de saída 0 se tudo passar, 1 caso contrário (útil em CI).
"""

import os
import platform
import shutil
import subprocess
import sys

falhas = 0
WINDOWS = platform.system() == "Windows"


def check(nome: str, condicao: bool, dica: str) -> None:
    global falhas
    print(f"[{'OK   ' if condicao else 'FALHA'}] {nome}")
    if not condicao:
        print(f"          -> {dica}")
        falhas += 1


def info(texto: str) -> None:
    print(f"[INFO ] {texto}")


print("=" * 62)
print("  Verificação do ambiente — Desenvolvimento de Projeto Web")
print("=" * 62)
info(f"Sistema: {platform.system()} {platform.release()}")
if WINDOWS:
    info("Guia de setup do Windows: docs/ambiente-setup-windows.md")
    info("Tabela de equivalências:  recursos/comandos-windows.md")

# --- Python ---------------------------------------------------------------
check(
    "Python >= 3.12",
    sys.version_info >= (3, 12),
    "Instale Python 3.12+ e recrie o ambiente virtual",
)
info(f"Versão detectada: {sys.version.split()[0]}")

check(
    "Ambiente virtual ativo",
    sys.prefix != sys.base_prefix,
    r".venv\Scripts\Activate.ps1" if WINDOWS else "source .venv/bin/activate",
)

# --- Git ------------------------------------------------------------------
git = shutil.which("git")
check("Git instalado", git is not None, "Instale o Git: https://git-scm.com/downloads")

if git:
    nome = subprocess.run(
        ["git", "config", "--global", "user.name"], capture_output=True, text=True
    ).stdout.strip()
    email = subprocess.run(
        ["git", "config", "--global", "user.email"], capture_output=True, text=True
    ).stdout.strip()
    check(
        "Git configurado (user.name e user.email)",
        bool(nome and email),
        'git config --global user.name "Seu Nome" && '
        'git config --global user.email "voce@exemplo.com"',
    )

# --- Django ---------------------------------------------------------------
try:
    import django

    check(
        "Django >= 5.0",
        django.VERSION >= (5, 0),
        "pip install 'django>=5.0,<6.0'",
    )
    info(f"Django {django.get_version()}")
except ImportError:
    check("Django instalado", False, "pip install 'django>=5.0,<6.0'")

# --- DRF e python-dotenv --------------------------------------------------
try:
    import rest_framework

    check("Django REST Framework", True, "")
    info(f"DRF {rest_framework.VERSION}")
except ImportError:
    check("Django REST Framework", False, "pip install djangorestframework")

try:
    import dotenv  # noqa: F401

    check("python-dotenv instalado", True, "")
except ImportError:
    check("python-dotenv instalado", False, "pip install python-dotenv")

# --- Node e pnpm (frontend) ----------------------------------------------
node = shutil.which("node")
check("Node.js instalado", node is not None, "Instale o Node 20 LTS (https://nodejs.org)")

if node:
    versao = subprocess.run(["node", "--version"], capture_output=True, text=True).stdout.strip()
    info(f"Node {versao}")
    try:
        maior = int(versao.lstrip("v").split(".")[0])
        check("Node >= 20", maior >= 20, "Atualize para o Node 20 LTS (use fnm ou nvm)")
    except ValueError:
        check("Node >= 20", False, "Não foi possível ler a versão do Node")

pnpm = shutil.which("pnpm")
check("pnpm instalado", pnpm is not None, "corepack enable; corepack prepare pnpm@latest --activate")
if pnpm:
    # `pnpm` é um .cmd no Windows: usar o caminho resolvido evita depender de shell=True
    versao_pnpm = subprocess.run([pnpm, "--version"], capture_output=True, text=True)
    info(f"pnpm {versao_pnpm.stdout.strip()}")

# --- Específico do Windows ------------------------------------------------
if WINDOWS:
    print("-" * 62)
    print("  Verificações específicas do Windows")
    print("-" * 62)

    # 1. curl.exe: no PowerShell, `curl` é alias de Invoke-WebRequest
    check(
        "curl.exe disponível",
        shutil.which("curl.exe") is not None or shutil.which("curl") is not None,
        "Use sempre 'curl.exe' no PowerShell — 'curl' é alias de Invoke-WebRequest",
    )

    # 2. Pasta do projeto fora do OneDrive, sem espaço e sem acento no caminho
    #    (causa nº 1 de lentidão, arquivo bloqueado e mudança fantasma no Git)
    cwd = os.getcwd()
    check(
        "Projeto fora do OneDrive",
        "onedrive" not in cwd.lower(),
        "Mova o projeto para C:\\dev — o OneDrive sincroniza .venv e node_modules, "
        "travando o pnpm install e produzindo mudanças fantasma no Git",
    )
    check(
        "Caminho sem espaço nem acento",
        " " not in cwd and cwd.isascii(),
        f"Caminho atual: {cwd} — mova para C:\\dev. Espaços e acentos quebram "
        "ferramentas que não põem aspas nos caminhos",
    )

    # 3. .gitattributes e autocrlf: evitam CRLF quebrando o deploy no M16
    raiz = None
    for pasta in (".", "..", "../.."):
        if os.path.isdir(os.path.join(pasta, ".git")):
            raiz = pasta
            break
    if raiz:
        check(
            ".gitattributes presente",
            os.path.isfile(os.path.join(raiz, ".gitattributes")),
            "Crie o .gitattributes com '*.sh text eol=lf' — sem ele o deploy falha "
            "com 'bad interpreter' (ver docs/ambiente-setup-windows.md, passo 11.2)",
        )
        autocrlf = subprocess.run(
            ["git", "config", "--get", "core.autocrlf"],
            capture_output=True, text=True,
        ).stdout.strip()
        check(
            "git core.autocrlf = input",
            autocrlf == "input",
            "git config --global core.autocrlf input",
        )
    else:
        info("Fora de um repositório Git — verificação de finais de linha pulada")

    # 4. requirements.txt em UTF-16: o `>` do PowerShell 5.1 grava assim, e o
    #    `pip install -r` falha com "Invalid requirement: 'ÿþd'" na máquina do colega.
    for pasta in (".", "backend"):
        req = os.path.join(pasta, "requirements.txt")
        if os.path.isfile(req):
            with open(req, "rb") as f:
                inicio = f.read(4)
            check(
                f"{req} em texto puro (não UTF-16)",
                not inicio.startswith((b"\xff\xfe", b"\xfe\xff")),
                "Gerado com '>' no PowerShell 5.1, que grava UTF-16. Refaça com: "
                "pip freeze | Out-File -FilePath requirements.txt -Encoding ascii",
            )
            break

    # 5. Avisos que não são verificáveis por script, mas custam caro quando ignorados
    info("Gunicorn não roda no Windows. No M16, use: pip install waitress")
    info('Variáveis inline: use $env:DEBUG="False"; python ... (e limpe depois)')
    info("Ao gerar arquivo de texto, use Out-File -Encoding ascii — nunca '>'")

# --- Docker (opcional até o M04) -----------------------------------------
if shutil.which("docker"):
    rodando = subprocess.run(["docker", "info"], capture_output=True).returncode == 0
    check(
        "Docker em execução",
        rodando,
        "Abra o Docker Desktop (opcional até o M04)",
    )
else:
    info("Docker não instalado — opcional até o M04")

# --- Resultado ------------------------------------------------------------
print("-" * 62)
if falhas == 0:
    print("Ambiente pronto. Bom curso!")
    if WINDOWS:
        print("Antes da primeira aula, leia as seis armadilhas do Windows:")
        print("  recursos/comandos-windows.md (seção 2)")
else:
    print(f"{falhas} item(ns) precisam de correção antes da aula.")
print("-" * 62)

sys.exit(0 if falhas == 0 else 1)

File: recursos/codigo/verifica-ambiente.mjs
#!/usr/bin/env node
/**
 * Verificação do ambiente da disciplina DPW.
 *
 * Este script CONFERE o ambiente; ele não instala nem configura nada. Montar o
 * ambiente é conteúdo da disciplina e os comandos são digitados por você — aqui
 * só entra o diagnóstico, com a dica de correção de cada falha.
 *
 * A instalação acontece em três momentos, e o script confere só o que já deveria
 * existir naquele ponto do curso:
 *
 *     node verifica-ambiente.mjs                 semana 1  (M00)
 *     node verifica-ambiente.mjs --etapa m03     antes do M03: + backend
 *     node verifica-ambiente.mjs --etapa m05     antes do M05: + Docker
 *
 * Código de saída 0 se tudo passar, 1 caso contrário (útil em CI).
 */

import { execSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { platform, release } from "node:os";

const WINDOWS = platform() === "win32";
const ETAPAS = ["m00", "m03", "m05"];

const i = process.argv.indexOf("--etapa");
const etapa = i !== -1 && process.argv[i + 1] ? process.argv[i + 1].toLowerCase() : "m00";
if (!ETAPAS.includes(etapa)) {
  console.error(`Etapa desconhecida: ${etapa}. Use uma de ${ETAPAS.join(", ")}.`);
  process.exit(2);
}
const EXIGE_BACKEND = ETAPAS.indexOf(etapa) >= ETAPAS.indexOf("m03");
const EXIGE_DOCKER = ETAPAS.indexOf(etapa) >= ETAPAS.indexOf("m05");

let falhas = 0;

const check = (nome, condicao, dica) => {
  console.log(`[${condicao ? "OK   " : "FALHA"}] ${nome}`);
  if (!condicao) {
    console.log(`          -> ${dica}`);
    falhas += 1;
  }
};
const info = (texto) => console.log(`[INFO ] ${texto}`);

/** Roda um comando e devolve a saída, ou null se o comando não existir/falhar. */
const rodar = (cmd) => {
  try {
    return execSync(cmd, { encoding: "utf8", stdio: ["ignore", "pipe", "ignore"] }).trim();
  } catch {
    return null;
  }
};

/** Extrai o "major" de uma string tipo "v20.11.0" ou "9.12.3". */
const maior = (versao) => Number.parseInt(String(versao).replace(/^v/, "").split(".")[0], 10);

console.log("=".repeat(62));
console.log("  Verificação do ambiente — Desenvolvimento de Projeto Web");
console.log("=".repeat(62));
info(`Sistema: ${platform()} ${release()}`);
info(`Etapa verificada: ${etapa.toUpperCase()}`);
if (!EXIGE_BACKEND) {
  info("Backend e Docker ainda nao sao exigidos — use --etapa m03 / m05 depois");
}
if (WINDOWS) {
  info("Guia de setup do Windows: docs/ambiente-setup-windows.md");
  info("Tabela de equivalencias:  recursos/comandos-windows.md");
}

// --- Node -----------------------------------------------------------------
const versaoNode = process.versions.node;
check(
  "Node >= 20",
  maior(versaoNode) >= 20,
  "Instale o Node 20 LTS (https://nodejs.org) e reabra o terminal",
);
info(`Node v${versaoNode}`);

// --- pnpm -----------------------------------------------------------------
const versaoPnpm = rodar("pnpm --version");
check(
  "pnpm instalado",
  versaoPnpm !== null,
  "corepack enable && corepack prepare pnpm@latest --activate",
);
if (versaoPnpm) {
  info(`pnpm ${versaoPnpm}`);
  check("pnpm >= 9", maior(versaoPnpm) >= 9, "corepack prepare pnpm@latest --activate");
}

// --- Git ------------------------------------------------------------------
const versaoGit = rodar("git --version");
check("Git instalado", versaoGit !== null, "Instale o Git: https://git-scm.com/downloads");

if (versaoGit) {
  const nome = rodar("git config --global user.name");
  const email = rodar("git config --global user.email");
  check(
    "Git configurado (user.name e user.email)",
    Boolean(nome && email),
    'git config --global user.name "Seu Nome" e ' +
      'git config --global user.email "voce@exemplo.com"',
  );
}

// --- Repositório ----------------------------------------------------------
let raiz = null;
for (const pasta of [".", "..", "../.."]) {
  if (existsSync(join(pasta, ".git"))) {
    raiz = pasta;
    break;
  }
}

if (raiz) {
  check(
    ".gitattributes presente",
    existsSync(join(raiz, ".gitattributes")),
    "Crie o .gitattributes com '*.sh text eol=lf' — sem ele o deploy do M16 falha " +
      "com 'bad interpreter' (ver ambiente-setup.md, secao 10)",
  );
} else {
  info("Fora de um repositorio Git — verificacoes do repositorio puladas");
}

// --- Backend (a partir do M03) -------------------------------------------
if (EXIGE_BACKEND) {
  let achou = false;
  for (const pasta of ["backend", "../backend", "."]) {
    const pkg = join(pasta, "package.json");
    if (existsSync(pkg)) {
      const deps = JSON.parse(readFileSync(pkg, "utf8")).dependencies ?? {};
      check(
        "NestJS instalado",
        "@nestjs/core" in deps,
        "Rode 'pnpm install' na raiz do monorepo",
      );
      check(
        "TypeORM instalado",
        "typeorm" in deps,
        "pnpm --filter backend add @nestjs/typeorm typeorm",
      );
      achou = true;
      break;
    }
  }
  if (!achou) {
    check(
      "package.json do backend encontrado",
      false,
      "Rode este script de dentro do monorepo (o M03 cria a pasta backend/)",
    );
  }
}

// --- Docker (a partir do M05) --------------------------------------------
if (EXIGE_DOCKER) {
  if (rodar("docker --version")) {
    check(
      "Docker em execucao",
      rodar("docker info") !== null,
      WINDOWS
        ? "Abra o Docker Desktop pelo menu Iniciar e espere a baleia estabilizar"
        : "sudo systemctl start docker (Linux) ou abra o Docker Desktop (macOS)",
    );
  } else {
    check(
      "Docker instalado",
      false,
      WINDOWS
        ? "No Windows exige o WSL2 antes — ver ambiente-setup-windows.md, passo 7"
        : "https://docs.docker.com/engine/install/",
    );
  }
}

// --- Específico do Windows ------------------------------------------------
if (WINDOWS) {
  console.log("-".repeat(62));
  console.log("  Verificacoes especificas do Windows");
  console.log("-".repeat(62));

  // 1. curl.exe: no PowerShell, `curl` e alias de Invoke-WebRequest
  check(
    "curl.exe disponivel",
    rodar("curl.exe --version") !== null,
    "Use sempre 'curl.exe' no PowerShell — 'curl' e alias de Invoke-WebRequest",
  );

  // 2. Pasta do projeto: OneDrive, espaco e acento sao a causa n1 de problema
  const cwd = process.cwd();
  check(
    "Projeto fora do OneDrive",
    !cwd.toLowerCase().includes("onedrive"),
    "Mova o projeto para C:\\dev — o OneDrive sincroniza node_modules, travando o " +
      "pnpm install e produzindo mudancas fantasma no Git",
  );
  check(
    "Caminho sem espaco nem acento",
    !cwd.includes(" ") && /^[\x20-\x7E]*$/.test(cwd.replace(/ /g, "")),
    `Caminho atual: ${cwd} — mova para C:\\dev. Espacos e acentos quebram ` +
      "ferramentas que nao poem aspas nos caminhos",
  );

  // 3. Finais de linha
  if (raiz) {
    check(
      "git core.autocrlf = input",
      rodar("git config --get core.autocrlf") === "input",
      "git config --global core.autocrlf input",
    );
  }

  // 4. Caminhos longos: um monorepo tem duas arvores de node_modules
  check(
    "core.longpaths habilitado",
    rodar("git config --get core.longpaths") === "true",
    "Num PowerShell como administrador: git config --system core.longpaths true",
  );

  info('Variaveis inline nao existem: use $env:VAR="x"; comando (e limpe depois)');
  info("Ao gerar arquivo de texto, use Out-File -Encoding utf8 — nunca '>'");
}

// --- Resultado ------------------------------------------------------------
console.log("-".repeat(62));
if (falhas === 0) {
  console.log(`Ambiente da etapa ${etapa.toUpperCase()} pronto. Bom curso!`);
  if (etapa === "m00") console.log("Antes do M03, rode de novo com: --etapa m03");
  else if (etapa === "m03") console.log("Antes do M05, rode de novo com: --etapa m05");
  if (WINDOWS) {
    console.log("Leia as cinco armadilhas do Windows: recursos/comandos-windows.md (secao 2)");
  }
} else {
  console.log(`${falhas} item(ns) precisam de correcao antes da aula.`);
}
console.log("-".repeat(62));

process.exit(falhas === 0 ? 0 : 1);
